storage/mysql/storage_test.go

```diff
@@ -552,6 +552,82 @@ func TestDequeueLeavesGuardInterval(t *testing.T) {
 	}
 }
 
+func TestDequeueLeavesTimeOrdering(t *testing.T) {
+	// Queue two small batches of leaves at different timestamps. Do two separate dequeue
+	// transactions and make sure the returned leaves are respecting the time ordering of the
+	// queue.
+
+	logID := createLogID("TestDequeueLeavesTimeOrdering")
+	db := prepareTestLogDB(logID, t)
+	defer db.Close()
+	s := prepareTestLogStorage(logID, t)
+	batchSize := 2
+
+	{
+		tx := beginLogTx(s, t)
+		defer failIfTXStillOpen(t, "TestDequeueLeavesTimeOrdering", tx)
+
+		leaves := createTestLeaves(int64(batchSize), 0)
+		leaves2 := createTestLeaves(int64(batchSize), int64(batchSize))
+
+		if err := tx.QueueLeaves(leaves, fakeQueueTime); err != nil {
+			t.Fatalf("Failed to queue leaves: %v", err)
+		}
+
+		// These are one second earlier so should be dequeued first
+		if err := tx.QueueLeaves(leaves2, fakeQueueTime.Add(-time.Second)); err != nil {
+			t.Fatalf("Failed to queue leaves2: %v", err)
+		}
+
+		commit(tx, t)
+	}
+
+	{
+		// Now try to dequeue two leaves and we should get the second batch
+		tx2 := beginLogTx(s, t)
+		defer failIfTXStillOpen(t, "TestDequeueLeavesTimeOrdering", tx2)
+		dequeue1, err := tx2.DequeueLeaves(batchSize, fakeQueueTime)
+
+		if err != nil {
+			t.Fatalf("Failed to dequeue leaves (1st): %v", err)
```

Comment:
Sorry, I meant:
   change to "DequeueLeaves(...) = %v ..." style messages (and elsewhere)
but I was too lazy to type it all out again :)